### PR TITLE
[lang] Implement experimental CG(Conjugate Gradient) solver in Taichi-lang

### PR DESCRIPTION
**Pull request by houkensjtu**
Issue: #7634 

### Brief Summary
This PR implements a matrix-free CG (Conjugate-Gradient) solver in Taichi. The solver targets to solve the linear equation system:

$$ Ax = b$$

where $A$ is implicitly represented as a `LinearOperator` instead of a explicitly stored matrix, hence the name "matrix-free".

### CODE REVIEW

File: python/taichi/linalg/taichi_cg.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    def __init__(self, matvec_kernel):
+        self._matvec = matvec_kernel
+
+    def matvec(self, x, Ax):
+        assert x.shape == Ax.shape, "Dimension mismatch x.shape != Ax.shape."
+        self._matvec(x, Ax)
+
+
+def taichi_cg_solver(A, b, x, tol=1e-6, maxiter=5000, quiet=True):
+    assert b.dtype == x.dtype, "Dtype mismatch b.dtype != x.dtype."
```
Comment (turbo0628) [inline python/taichi/linalg/taichi_cg.py:17]: I think you should raise an exception here :)

File: python/taichi/linalg/taichi_cg.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    if str(b.dtype) == 'f32':
+        solver_dtype = ti.f32
+    elif str(b.dtype) == 'f64':
+        solver_dtype = ti.f64
+    else:
+        print("Not supported dtype:", b.dtype)
```
Comment (turbo0628) [inline python/taichi/linalg/taichi_cg.py:23]: ditto

File: python/taichi/linalg/taichi_cg.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        exit()
+
+    assert b.shape == x.shape, "Dimension mismatch b.shape != x.shape."
```
Comment (turbo0628) [inline python/taichi/linalg/taichi_cg.py:26]: ditto

File: python/taichi/linalg/taichi_cg.py (hunk 1 above)
Comment (houkensjtu) [reply python/taichi/linalg/taichi_cg.py]: Thank you for the suggestion!

File: python/taichi/linalg/taichi_cg.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+@ti.data_oriented
+class LinearOperator:
+    def __init__(self, matvec_kernel):
+        self._matvec = matvec_kernel
+
+    def matvec(self, x, Ax):
+        if x.shape != Ax.shape:
+            raise TaichiRuntimeError("Dimension mismatch x.shape != Ax.shape.")
```
Comment (FantasyVR) [inline python/taichi/linalg/taichi_cg.py:15]:
```suggestion
            raise TaichiRuntimeError(f"Dimension mismatch {x.shape} != {Ax.shape}.")
```

File: python/taichi/linalg/taichi_cg.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        self._matvec(x, Ax)
+
+
+def taichi_cg_solver(A, b, x, tol=1e-6, maxiter=5000, quiet=True):
+    if b.dtype != x.dtype:
+        raise TaichiTypeError("Dtype mismatch b.dtype != x.dtype.")
```
Comment (FantasyVR) [inline python/taichi/linalg/taichi_cg.py:21]: ditto

File: python/taichi/linalg/taichi_cg.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if str(b.dtype) == 'f32':
+        solver_dtype = ti.f32
+    elif str(b.dtype) == 'f64':
+        solver_dtype = ti.f64
+    else:
+        raise TaichiTypeError(f"Not supported dtype:{b.dtype}")
+    if b.shape != x.shape:
+        raise TaichiRuntimeError("Dimension mismatch b.shape != x.shape.")
```
Comment (FantasyVR) [inline python/taichi/linalg/taichi_cg.py:29]:
```suggestion
        raise TaichiRuntimeError("Dimension mismatch {b.shape} != {x.shape}.")
```

**Review by FantasyVR — COMMENTED** [review]
Thanks

**Review by FantasyVR — APPROVED** [review]

**Review by turbo0628 — APPROVED** [review]
LGTM!

**Pull request merged**